src/platform/audio/ffmpegaudiofilecache.cpp

```diff
@@ -0,0 +1,537 @@
+
+    // Try to find existing data for this file to avoid loading it if unneeded.
+    auto it = m_cacheMap.find(filename);
+
+    if (it != m_cacheMap.end()) {
+        ++it->second.ref_count;
+
+        return it->second.wave_data;
+    }
+
+    // Load the file from disk
+    File *file = g_theFileSystem->Open(filename, File::READ | File::BINARY | File::BUFFERED);
+
+    if (file == nullptr) {
+        if (filename.Is_Not_Empty()) {
+            captainslog_warn("Missing audio file '%s', could not cache.", filename.Str());
+        }
+
+        return nullptr;
+    }
+
+    FFmpegOpenAudioFile open_audio;
+    open_audio.wave_data = static_cast<uint8_t *>(av_malloc(sizeof(WavHeader)));
+    open_audio.data_size = sizeof(WavHeader);
+
+    if (!Open_FFmpeg_Contexts(&open_audio, file)) {
+        captainslog_warn("Failed to load audio file '%s', could not cache.", filename.Str());
+        Release_Open_Audio(&open_audio);
+        file->Close();
+        return nullptr;
+    }
+
+    if (!Decode_FFmpeg(&open_audio)) {
+        captainslog_warn("Failed to decode audio file '%s', could not cache.", filename.Str());
+        Close_FFmpeg_Contexts(&open_audio);
+        Release_Open_Audio(&open_audio);
+        file->Close();
+        return nullptr;
+    }
+
+    Fill_Wave_Data(&open_audio);
+    Close_FFmpeg_Contexts(&open_audio);
+    file->Close();
+
+    open_audio.ref_count = 1;
+    m_currentSize += open_audio.data_size;
+
+    // m_maxSize prevents using overly large amounts of memory, so if we are over it, unload some other samples.
+    if (m_currentSize > m_maxSize && !Free_Space_For_Sample(open_audio)) {
+        captainslog_warn("Cannot play audio file since cache is full: %s", filename.Str());
+        m_currentSize -= open_audio.data_size;
+        Release_Open_Audio(&open_audio);
+
+        return nullptr;
+    }
+
+    m_cacheMap[filename] = open_audio;
+
+    return open_audio.wave_data;
+}
+/**
+ * Opens an audio file for an event. Reads from the cache if available or loads from file if not.
+ */
+void *FFmpegAudioFileCache::Open_File(AudioEventRTS *audio_event)
+{
+    ScopedMutexClass lock(&m_mutex);
+    Utf8String filename;
+
+    // What part of an event are we playing?
+    switch (audio_event->Get_Next_Play_Portion()) {
+        case 0:
+            filename = audio_event->Get_Attack_Name();
+            break;
+        case 1:
+            filename = audio_event->Get_File_Name();
+            break;
+        case 2:
+            filename = audio_event->Get_Decay_Name();
+            break;
+        case 3:
+            return nullptr;
+        default:
+            break;
+    }
+
+    captainslog_trace("FFmpegAudioFileCache: opening file %s", filename.Str());
+
+    // Try to find existing data for this file to avoid loading it if unneeded.
+    auto it = m_cacheMap.find(filename);
+
+    if (it != m_cacheMap.end()) {
+        ++it->second.ref_count;
```

Comment:
Minor: `++(it->second.ref_count);` will make it easier to read, as one could think on first sight that it is `it` being incremented here.